pecos/core/utils/mmap_util.hpp

```diff
@@ -0,0 +1,530 @@
+            if (!fp) {
+                throw std::runtime_error("Load metadata: Open file failed.");
+            }
+            metadata_.load(fp);
+            if (fclose(fp) != 0) {
+                throw std::runtime_error("Load metadata: Close file failed.");
+            }
+
+            // Load mmap
+            int fd = open(file_name.c_str(), O_RDONLY);
+            if (fd == -1) {
+                throw std::runtime_error("Load Mmap file: Open file failed.");
+            }
+            load_mmap_(fd, pre_load);
+            if (close(fd) < 0) {
+                throw std::runtime_error("Load Mmap file: Close file failed.");
+            }
+        }
+        /* Destructor frees the memory-mapped region */
+        ~MmapStoreLoad_() { free_mmap_(); }
+
+        /* Load a simple-serializable vector(array)
+         * Parameters:
+         *      n_elements (uint64_t): Number of elements to load for the array.
+         */
+        template<class T, class TT=T, if_simple_serializable<TT> = true>
+        T* fget_multiple(uint64_t n_elements) {
+            auto meta_info = metadata_.next();
+
+            // Verification
+            if (n_elements * static_cast<uint64_t>(sizeof(T)) != meta_info.size) {
+                throw std::runtime_error(
+                    "This block contains " + std::to_string(meta_info.size) +
+                    " bytes data, retrieving " + std::to_string(n_elements * static_cast<uint64_t>(sizeof(T))) +
+                    " bytes not equal. Please double check.");
+            }
+
+            return reinterpret_cast< T* >(reinterpret_cast< char * >(mmap_ptr_) + meta_info.offset);
+        }
+
+        /* Load one element */
+        template<class T, class TT=T, if_simple_serializable<TT> = true>
+        T fget_one() { // Call fget_multiple
+            return *fget_multiple<T>(1u);
+        }
+
+    private:
+        MmapMetadata_ metadata_; // Metadata
+        void* mmap_ptr_ = nullptr; // Memory-mapped pointer
+        uint64_t mmap_size_ = 0;// Memory-mapped file size
+
+        /* Create memory-mapped region */
+        void load_mmap_(const int fd, const bool pre_load) {
+            // Get file size
+            struct stat file_stat;
+            fstat(fd, &file_stat);
+            mmap_size_ = file_stat.st_size;
+            if (mmap_size_ <= 0) {
+                throw std::runtime_error("Memory mapped file size should be positive, got: " + std::to_string(mmap_size_));
+            }
+
+            // Creat mmap
+            int mmap_flags = MAP_SHARED;
+            if (pre_load) { // pre-fault all pages to load them into memory
+                mmap_flags |= MAP_POPULATE;
+            }
+            mmap_ptr_ = mmap(NULL, mmap_size_, PROT_READ, mmap_flags, fd, 0);
+            if (mmap_ptr_ == MAP_FAILED) {
+                throw std::runtime_error("Memory map failed.");
+            }
+        }
+
+        /* Free memory-mapped region */
+        void free_mmap_() {
```

Comment:
If we don't plan to reveal this method and only use it once in the destructor, I would suggest to move the content into the destructor itself.